### PR TITLE
Add Terastallization support

### PR DESCRIPTION
**Pull request by ajhyndman**
- Adds teraType to Pokemon class
- Accounts for teraType in STAB calculation
- Accounts for teraType in resistance calculation

You might want to double-check especially the resistance math.  I am not confident I captured all edge cases there.

### CODE REVIEW

**Comment by ajhyndman** [comment]
It looks like I broke a couple tests.  I'm investigating.

**Comment by ajhyndman** [comment]
Okay, think I got it.  I didn't recall that iron ball makes effectiveness exactly 1x regardless of second type!

**Comment by ajhyndman** [comment]
Closed in favor of #484

**Comment by ajhyndman** [comment]
@AustinXII I think I got tests passing whenever you have a chance to take a look!

File: calc/src/mechanics/gen78.ts [hunk 1]
```diff
@@ -230,26 +235,48 @@ export function calculateSMSS(
...
+  );
   const type2Effectiveness = defender.types[1]
-    ? getMoveEffectiveness(gen, move, defender.types[1], isGhostRevealed, field.isGravity)
+    ? getMoveEffectiveness(
+      gen,
+      move,
+      defender.types[1],
+      isGhostRevealed,
+      field.isGravity,
+      isRingTarget
```
Comment (KrisXV) [inline calc/src/mechanics/gen78.ts:257]:
```suggestion
        gen,
        move,
        defender.types[1],
        isGhostRevealed,
        field.isGravity,
        isRingTarget
```

File: calc/src/mechanics/gen78.ts [hunk 2]
```diff
@@ -455,16 +482,18 @@ export function calculateSMSS(
...
-    } else {
-      stabMod = 6144;
-    }
+    stabMod += 2048;
   } else if (attacker.hasAbility('Protean', 'Libero')) {
-    stabMod = 6144;
+    stabMod += 2048;
     desc.attackerAbility = attacker.ability;
   }
+  if (attacker.teraType === move.type) {
```
Comment (KrisXV) [inline calc/src/mechanics/gen78.ts:490]: The stabBoost only gets an additional 4/3 multiplier if the base type of the Pokemon matches the tera type

File: calc/src/mechanics/gen78.ts [hunk 3]
```diff
@@ -388,7 +415,7 @@ export function calculateSMSS(
   // #region (Special) Attack
   const attack = calculateAttackSMSS(gen, attacker, defender, move, field, desc, isCritical);
   const attackSource = move.named('Foul Play') ? defender : attacker;
-  if (move.named('Photon Geyser', 'Light That Burns The Sky')) {
+  if (move.named('Photon Geyser', 'Light That Burns The Sky', 'Tera Blast')) {
```
Comment (KrisXV) [inline calc/src/mechanics/gen78.ts:418]: tera blast only does the category adjustment if the user is terastallized

File: calc/src/mechanics/gen78.ts (hunk 1 above)
Comment (ajhyndman) [reply calc/src/mechanics/gen78.ts:257]: The lint rule configured on this repo actually rejects that indentation.

Comment (KrisXV) [reply calc/src/mechanics/gen78.ts:257]: That's weird, it shouldn't be on the same indentation as the first line

File: calc/src/mechanics/gen78.ts (hunk 2 above)
Comment (ajhyndman) [reply calc/src/mechanics/gen78.ts:490]: That's true, but I've organized this so that normal STAB and tera STAB stack additively, handling both cases.

If you check the math, I think you'll find it works out.

File: calc/src/mechanics/gen78.ts (hunk 3 above)
Comment (ajhyndman) [reply calc/src/mechanics/gen78.ts:418]: Whoah, I didn't actually know that.  Thanks for the catch!

File: calc/src/mechanics/gen78.ts (hunk 1 above)
Comment (ajhyndman) [reply calc/src/mechanics/gen78.ts:257]: Yeah, it looked wrong to me too.  🤷   You could always switch to prettier!

**Comment by AnnikaCodes** [comment]
Thanks!
